File: 2-introduction-to-deep-agents/typescript/m1/m1.8_homework_filled.ts
// typescript/m1/m1.8_homework_filled.ts
/**
 * Reference copy of m1.8_homework.ts with TODOs 1 and 2 filled in so you
 * can run it end to end and see what "done" looks like. This is just one
 * possible answer, so yours might be different. Explore!
 */

import { createInterface } from "node:readline/promises";
import { z } from "zod";

import { tool, ToolMessage, type InterruptOnConfig } from "langchain";
import { Command, INTERRUPT, isInterrupted, MemorySaver } from "@langchain/langgraph";
import { createDeepAgent } from "deepagents";

import { model } from "../models.js";

interface ActionRequest {
  name: string;
  args: Record<string, unknown>;
}

interface ApprovalRequest {
  actionRequests: ActionRequest[];
}

// TODO 1 filled in
const postTweet = tool(
  ({ content }: { content: string }) => `Tweet posted: ${JSON.stringify(content)}`,
  {
    name: "post_tweet",
    description: "Post a tweet with the given content.",
    schema: z.object({ content: z.string() }),
  }
);

// TODO 2 filled in
const SYSTEM_PROMPT = `You are a social media assistant that drafts and posts tweets.

Rules:
- Use post_tweet when the user asks you to post something.
- Keep tweets under 280 characters and upbeat in tone.
- Do not claim a tweet was posted until the tool result confirms it.
`;
const INITIAL_REQUEST = "Post a tweet announcing that our new product launches this Friday.";
const INTERRUPT_ON = {
  post_tweet: { allowedDecisions: ["approve", "edit", "reject"] },
} satisfies Record<string, boolean | InterruptOnConfig>;

const agent = createDeepAgent({
  model,
  tools: [postTweet],
  systemPrompt: SYSTEM_PROMPT,
  interruptOn: INTERRUPT_ON,
  checkpointer: new MemorySaver(),
});

const config = { configurable: { thread_id: "m1-8-homework-demo" } };

let result = await agent.invoke(
  { messages: [{ role: "user", content: INITIAL_REQUEST }] },
  config
);

const rl = createInterface({ input: process.stdin, output: process.stdout });

while (isInterrupted<ApprovalRequest>(result) && result[INTERRUPT].length) {
  const pending = result[INTERRUPT][0].value!;
  const decisions = [];

  for (const req of pending.actionRequests) {
    console.log(`\nApproval required for ${req.name}:`);
    console.log(req.args);

    const choice = (
      await rl.question("\nApprove, edit, or reject? (approve/edit/reject): ")
    )
      .trim()
      .toLowerCase();

    if (["approve", "yes", "y"].includes(choice)) {
      decisions.push({ type: "approve" });
    } else if (["edit", "e"].includes(choice)) {
      const editedArgs = { ...req.args };
      editedArgs.content = await rl.question("New tweet content: ");
      decisions.push({
        type: "edit",
        editedAction: { name: req.name, args: editedArgs },
      });
    } else {
      decisions.push({ type: "reject", message: "User rejected this tweet draft." });
    }
  }

  result = await agent.invoke(new Command({ resume: { decisions } }), config);
}

rl.close();
for (const msg of result.messages) {
  if (ToolMessage.isInstance(msg) && msg.name === "post_tweet") {
    console.log(msg.content);
    break;
  }
}

File: 2-introduction-to-deep-agents/python/m1/m1.8_homework_filled.py
# python/m1/m1.8_homework_filled.py
"""Reference copy of m1.8_homework.py with TODOs 1 and 2 filled in so you
can run it end to end and see what "done" looks like. This is just one
possible answer, so yours might be different. Explore!"""

import warnings

warnings.filterwarnings("ignore", category=DeprecationWarning)

from deepagents import create_deep_agent
from langchain_core.tools import tool
from langgraph.checkpoint.memory import MemorySaver
from langgraph.types import Command

from models import model


# TODO 1 filled in
@tool
def post_tweet(content: str) -> str:
    """Post a tweet with the given content."""
    return f"Tweet posted: {content!r}"


# TODO 2 filled in
SYSTEM_PROMPT = """You are a social media assistant that drafts and posts tweets.

Rules:
- Use post_tweet when the user asks you to post something.
- Keep tweets under 280 characters and upbeat in tone.
- Do not claim a tweet was posted until the tool result confirms it.
"""
INITIAL_REQUEST = "Post a tweet announcing that our new product launches this Friday."
INTERRUPT_ON = {"post_tweet": {"allowed_decisions": ["approve", "edit", "reject"]}}

agent = create_deep_agent(
    model=model,
    tools=[post_tweet],
    system_prompt=SYSTEM_PROMPT,
    interrupt_on=INTERRUPT_ON,
    checkpointer=MemorySaver(),
)

config = {"configurable": {"thread_id": "m1-8-homework-demo"}}

result = agent.invoke(
    {"messages": [{"role": "user", "content": INITIAL_REQUEST}]},
    config=config,
    version="v2",
)

while result.interrupts:
    pending = result.interrupts[0].value
    decisions = []
    for req in pending["action_requests"]:
        print(f"\nApproval required for {req['name']}:")
        print(req["args"])

        choice = input("\nApprove, edit, or reject? (approve/edit/reject): ").strip().lower()
        if choice in ("approve", "yes", "y"):
            decisions.append({"type": "approve"})
        elif choice in ("edit", "e"):
            edited_args = dict(req["args"])
            edited_args["content"] = input("New tweet content: ")
            decisions.append(
                {
                    "type": "edit",
                    "edited_action": {"name": req["name"], "args": edited_args},
                }
            )
        else:
            decisions.append(
                {"type": "reject", "message": "User rejected this tweet draft."}
            )

    result = agent.invoke(Command(resume={"decisions": decisions}), config=config, version="v2")

for msg in result.value["messages"]:
    if hasattr(msg, "name") and msg.name == "post_tweet":
        print(msg.content)
        break
